File: benchmarks/big-decimal/cell.mjs
// one (op, prec) cell in a fresh process. usage: node _cell.mjs <SRC|bundlePath> <op> <prec> [budgetMs]
const [src, op, precS, ms] = [process.argv[2], process.argv[3], +process.argv[4], +(process.argv[5]||400)];
const { ComputeEngine } = src === 'SRC'
  ? await import('../../src/compute-engine/index.ts') : await import(src);
const OPS = {
  ln:(c)=>['Ln',c+2], exp:(c)=>['Exp',['Divide',c+1,c+3]],
  sin:(c)=>['Sin',['Divide',c+1,c+3]], cos:(c)=>['Cos',['Divide',c+1,c+3]],
  tan:(c)=>['Tan',['Divide',c+1,c+3]], atan:(c)=>['Arctan',c+2],
  asin:(c)=>['Arcsin',['Divide',c+1,c+3]], sqrt:(c)=>['Sqrt',c+2],
};
const ce=new ComputeEngine(); ce.precision=precS; const b=OPS[op]; let c=0;
for(let i=0;i<5;i++) ce.box(b(c++)).N();
let n=0; const t0=process.hrtime.bigint(); let e=0n; const L=BigInt(ms)*1000000n;
do{ce.box(b(c++)).N();n++;if((n&3)===0)e=process.hrtime.bigint()-t0;}while(e<L);
console.log((Number(process.hrtime.bigint()-t0)/n/1e6).toFixed(3));

File: benchmarks/big-decimal/cell.py
# one (impl, op, prec) cell. usage: python _cell.py <sympy|mpmath> <op> <prec> [budgetMs]
import sys, time, sympy as sp, mpmath
impl, op, p = sys.argv[1], sys.argv[2], int(sys.argv[3])
ms = int(sys.argv[4]) if len(sys.argv)>4 else 400
SY={'ln':lambda c:sp.log(c+2).evalf(p),'exp':lambda c:sp.exp(sp.Rational(c+1,c+3)).evalf(p),
    'sin':lambda c:sp.sin(sp.Rational(c+1,c+3)).evalf(p),'cos':lambda c:sp.cos(sp.Rational(c+1,c+3)).evalf(p),
    'tan':lambda c:sp.tan(sp.Rational(c+1,c+3)).evalf(p),'atan':lambda c:sp.atan(c+2).evalf(p),
    'asin':lambda c:sp.asin(sp.Rational(c+1,c+3)).evalf(p),'sqrt':lambda c:sp.sqrt(c+2).evalf(p)}
def mp(c,fn):
    mpmath.mp.dps=p; return fn(c)
MP={'ln':lambda c:mp(c,lambda c:mpmath.log(c+2)),'exp':lambda c:mp(c,lambda c:mpmath.exp(mpmath.mpf(c+1)/(c+3))),
    'sin':lambda c:mp(c,lambda c:mpmath.sin(mpmath.mpf(c+1)/(c+3))),'cos':lambda c:mp(c,lambda c:mpmath.cos(mpmath.mpf(c+1)/(c+3))),
    'tan':lambda c:mp(c,lambda c:mpmath.tan(mpmath.mpf(c+1)/(c+3))),'atan':lambda c:mp(c,lambda c:mpmath.atan(c+2)),
    'asin':lambda c:mp(c,lambda c:mpmath.asin(mpmath.mpf(c+1)/(c+3))),'sqrt':lambda c:mp(c,lambda c:mpmath.sqrt(c+2))}
f=(SY if impl=='sympy' else MP)[op]
c=0
for _ in range(5): f(c); c+=1
n=0; t0=time.perf_counter()
while True:
    f(c); c+=1; n+=1
    if (n&3)==0 and time.perf_counter()-t0>=ms/1000: break
print(f"{(time.perf_counter()-t0)/n*1000:.3f}")
